### PR TITLE
Some tests cannot pass from 7AM to 8AM #4814

### LINKED ISSUE #4814: Some tests cannot pass from 7AM to 8AM
InstructorFeedbacksPageDataTest
InstructorFeedbackEditCopyUiTest
InstructorFeedbackEditPageUiTest
InstructorFeedbackPageUiTest

If we run those tests from 7AM to 8AM (i.e. from 11PM to 12AM in UTC), they always fail because the start time in FeedbackSessionsForm is adjusted to the next hour so the date in UTC timezone is also increased by 1. Consequently, all test check for the current date will fail.

### PR DESCRIPTION
**Pull request by YongJieYongJie**
Fixes #4814

### CODE REVIEW

**Comment by YongJieYongJie** [comment]
`InstructorFeedbacksPageDataTest` is fixed by simply using `TimeHelper.getNextHour()` instead of `now` to generate expected result

the other three are fixed by modifying `replaceUnpredictableValuesWithPlaceholders()` to address edge case by replacing (in actual page) `TimeHelper.formatDate(TimeHelper.getNextHour())` with `${today}`

// I had to merge the application of GodMode from Jun Hao's branch 4979 to pass the  `InstructorFeedbackPageUiTest`

File: src/test/resources/pages/instructorFeedbackCopyTrimmedSuccess.html [hunk 1]
```diff
@@ -746,7 +746,7 @@ <h5 class="col-xs-2" data-original-title="Select this option to use a custom tim
   <div aria-hidden="true" aria-labelledby="fsCopyModal" class="modal fade" id="fsCopyModal" role="dialog" tabindex="-1">
     <div class="modal-dialog">
       <div class="modal-content">
-        <form action="/page/instructorFeedbackEditCopy" method="post" name="form_copy_list" role="form">
+        <form action="/page/instructorFeedbackEditCopy" id="instructorCopyModalForm" method="post" role="form">
```
Comment (wkurniawan07) [inline src/test/resources/pages/instructorFeedbackCopyTrimmedSuccess.html]: I appreciate you trying to make all tests pass, but it remains that these changes shouldn't be part of this PR. Can you revert them?

Comment (junhaoyap) [inline src/test/resources/pages/instructorFeedbackCopyTrimmedSuccess.html]: It's odd that this is still being diff-ed considering that the branch that this is picked from has been merged

Comment (wkurniawan07) [inline src/test/resources/pages/instructorFeedbackCopyTrimmedSuccess.html]: Now that #4980 is merged (thanks @junhaoyap!) you can just merge it here.

**Comment by wkurniawan07** [comment]
Not that I disagree with the solution, but the moment you add another "replace" line to the algorithm, that will be applied to all HTML verification without any exception, and I prefer to avoid it unless it's a must (in this case, there's an alternative of simply moving the times forward by one hour).
Either way you choose, there are more things need to be done:
1. If you choose to scrap it (and move the times used by one or more hours as suggested), you need to fix the affected tests.
2. If you choose to keep it, add a test for it in `HtmlHelperTest.testReplacement`. In particular, add `<!-- nexthour.date -->` in `godmode.html` and a replacement line in `HtmlHelper.injectContextDependentValuesForTest`, and fix whatever needs fix in `HtmlHelperTest` afterwards.

File: src/test/resources/pages/instructorFeedbackCopyTrimmedSuccess.html (hunk 1 above)
Comment (wkurniawan07) [inline src/test/resources/pages/instructorFeedbackCopyTrimmedSuccess.html]: I don't think it's a merge; it's a cherry-pick, so the diff stays.

**Comment by YongJieYongJie** [comment]
I just merged the master back in, the GodMode diff in now gone.
Will look into the comments later today.

**Comment by YongJieYongJie** [comment]
@wkurniawan07 
Thanks so much for the detailed explanation.

You are right, my solution (i.e., additional conditional replacement) will cause problem if we ever add a placeholder for `${tomorrow}` because there is no way of distinguishing `${tomorrow}` from `${nexthour.date)` between UTC 2300-0000

The **root problem is** that we are **doing replacement in UTC time for a page that is rendered in the user's timezone**. So instead of:
`.replace(TimeHelper.formatDate(now), "${today}")`,
we should be using something like:
`.replace(todayInUserTimeZone, "${today}")`.

(Also, I think shifting the time forward for one hour will only pass the problem to another timezone.)

**Comment by wkurniawan07** [comment]
@YongJieYongJie I don't think that's a wise thing to do. Just shift the time on the json data source.

**Comment by YongJieYongJie** [comment]
@wkurniawan07 for at least the `InstructorFeedbackEditCopyUiTest`, it's not possible to solve the issue by changing the JSON because the actual date generated is based on system time:

after clicking copy:
image

the date generated is based on system time `.getNextHour()`:
image

I also realised if the action is performed before UTC 2300, the wrong date (i.e., yesterday's date) is shown and the test is showing **false negative**.

(I'll be digging into more details.)

**Comment by wkurniawan07** [comment]
@YongJieYongJie I see what you mean. Take your time to understand what's going on, meanwhile I'll do some investigation too myself.
This issue was never on my radar because it only happens 1 hour a day (and who in their right mind wakes up on 7AM to run tests? :p) but it's good if it can be fixed.

**Comment by YongJieYongJie** [comment]
`getNextHour()` is only called at line 140 of [FeedbackSessionsForm.java][fsf]:

> `newFsForm.fsStartDate = feedbackSession == null ? TimeHelper.formatDate(TimeHelper.getNextHour()) : TimeHelper.formatDate(feedbackSession.startTime);`

I'll change the second line to take into account user's time zone.
However, right now the selected time zone is not easily accessible, so I will to extract a method to get the currently selected timezone from the HTML option tags in `newFsForm.timezoneSelectField` by seeing which of the tag is `selected`:

> `newFsForm.timezoneSelectField = PageData.getTimeZoneOptionsAsElementTags(feedbackSession == null ? Const.DOUBLE_UNINITIALIZED : feedbackSession.timeZone);`

[fsf]: https://github.com/TEAMMATES/teammates/blob/master/src/main/java/teammates/ui/template/FeedbackSessionsForm.java#L140

**Comment by YongJieYongJie** [comment]
Updates:

Method discussed above **not feasible** because server has no knowledge of user's timezone. TEAMMATES's current design is to use JavaScript to handle timezone detection: the methods in `instructorFeedback.js`, such as `isTimeZoneIntialized()`

**Comment by YongJieYongJie** [comment]
@wkurniawan07 
Perhaps we can put fixing the tests on hold first, because these tests are dependent on the date bug. I'll post a new issue to solve that bug first.
(date bug: if you visit new feedback session page as an instructor before 7am UTC+8, yesterday's date will be shown, and if you visit 4pm UTC-8, tomorrow's date will be shown, etc.)

Also, `InstructorFeedbacksPageDataTest` can also be fixed in a separate issue because it is failing for a different reason from the other three tests.
(The other 3 are failing at `verifyHtmlMainContent()` due to the incongruity between user timezone in actual html and UTC in expected html in `HtmlHelper.java`)

**Comment by wkurniawan07** [comment]
Noted, thanks for all the work :+1:

**Comment by YongJieYongJie** [comment]
`AdminHomePageUiTest` fails when run on UTC+8 from 7am to 8am too

**Comment by YongJieYongJie** [comment]
will be fixing ##5165 first

**Comment by wkurniawan07** [comment]
@YongJieYongJie any updates here? I was wondering when I can remove than "on hold" label and get this back up again.

**Comment by YongJieYongJie** [comment]
@wkurniawan07 
actually I was done a few days ago but wanted to wait for two other PRs to really be merged before notifying you, I'll merge the new conflicts by tonight and notify you when ready

**Comment by YongJieYongJie** [comment]
@wkurniawan07 
Ready for review!

Here's what I did:
I did a `grep` of all `.html` expectations containing `${today}`, then I did a `grep` of all `.java` tests using those `.html` expectations. I found all all references to `${today}` actually meant `${date.nexthour}`, so I made 2 changes:
1. change the name from `today` to `date.nexthour`
2. use `getNextHour()` instead of just `now` when doing replacement in `HtmlHelper.java`

`InstructorFeedbacksPageDataTest.java` was failing for a different and simpler reason from the other tests and only requires using `getNextHour()` within the test itself

// here's a picture of the `grep`, hopefully I got it right
image

**Comment by wkurniawan07** [comment]
@YongJieYongJie you know, you could've used Eclipse's project-wide search to do the same :p

**Comment by YongJieYongJie** [comment]
@wkurniawan07 Eclipse is a little bit slow on my aging computer =(
Shall try tweaking the Eclipse launch config file soon

**Comment by wkurniawan07** [comment]
@YongJieYongJie all seems well, but can you do one last check that the problem isn't "shifted", i.e. perhaps now it fails from 6AM to 7AM?

**Comment by YongJieYongJie** [comment]
@wkurniawan07 I checked from 6AM to 7AM and also 8AM to 9AM, all tests pass except for `InstructorCommentsPageUiTest` at line 167, as mention in #5204
